rust/frontend/src/value.rs

```diff
@@ -0,0 +1,540 @@
+//! This module implements [`TVMArgValue`] and [`TVMRetValue`] types and their conversions
+//! to other support TVMValue.
+//!
+//! # Example
+//!
+//! ```
+//! let a = 42i8;
+//! let b = TVMArgValue::from(&a);
+//! assert_eq!(b.to_int() as i8, a);
+//! ```
+
+use std::ffi::{CStr, CString};
+use std::fmt::{self, Debug, Formatter};
+use std::marker::PhantomData;
+use std::mem;
+use std::ops::{Deref, DerefMut};
+use std::os::raw::{c_char, c_void};
+
+use ts;
+
+use ty::TypeCode;
+use Function;
+use Module;
+use NDArray;
+use TVMByteArray;
+use TVMContext;
+use TVMDeviceType;
+use TVMType;
+
+#[derive(Debug, Clone, PartialOrd, Ord, PartialEq, Eq)]
+pub(crate) enum ValueKind {
+    Int,
+    Float,
+    Handle,
+    Str,
+    Bytes,
+    Type,
+    Context,
+    Return,
+}
+
+/// Wrapper around underlying TVMValue.
+#[derive(Clone)]
+pub struct TVMValue {
+    pub(crate) kind: ValueKind,
+    pub(crate) inner: ts::TVMValue,
+}
+
+impl TVMValue {
+    pub(crate) fn new(kind: ValueKind, inner: ts::TVMValue) -> Self {
+        TVMValue { kind, inner }
+    }
+
+    pub fn to_int(&self) -> i64 {
+        unsafe { self.inner.v_int64 }
+    }
+}
+
+macro_rules! impl_prim_val {
+    ($type:ty, $kind:expr, $field:ident, $cast:ty) => {
+        impl<'a> From<&'a $type> for TVMValue {
+            fn from(arg: &$type) -> Self {
+                let inner = ts::TVMValue {
+                    $field: *arg as $cast,
+                };
+                Self::new($kind, inner)
+            }
+        }
+
+        impl<'a> From<&'a mut $type> for TVMValue {
+            fn from(arg: &mut $type) -> Self {
+                let inner = ts::TVMValue {
+                    $field: *arg as $cast,
+                };
+                Self::new($kind, inner)
+            }
+        }
+    };
+}
+
+impl_prim_val!(usize, ValueKind::Int, v_int64, i64);
+impl_prim_val!(i64, ValueKind::Int, v_int64, i64);
+impl_prim_val!(i32, ValueKind::Int, v_int64, i64);
+impl_prim_val!(i16, ValueKind::Int, v_int64, i64);
+impl_prim_val!(i8, ValueKind::Int, v_int64, i64);
+impl_prim_val!(u64, ValueKind::Int, v_int64, i64);
+impl_prim_val!(u32, ValueKind::Int, v_int64, i64);
+impl_prim_val!(u16, ValueKind::Int, v_int64, i64);
+impl_prim_val!(u8, ValueKind::Int, v_int64, i64);
+impl_prim_val!(bool, ValueKind::Int, v_int64, i64);
+impl_prim_val!(f64, ValueKind::Float, v_float64, f64);
+impl_prim_val!(f32, ValueKind::Float, v_float64, f64);
+
+impl<'a> From<&'a [u8]> for TVMValue {
+    fn from(arg: &[u8]) -> TVMValue {
+        let len = arg.len();
+        let arg = CString::new(arg).unwrap();
```

Comment:
please refer to the earlier comment about converting via `CString`

Reply:
As I'm resolving your comments, using `error_chain` has been very help thus far. However, in here and the other similar conversions to `TVMValue`

1. `CString::new` is *necessary*, as Rust's first need to own it and check for any 0 bytes so that it can work through FFI by passing the ownership to C and `mem::forget` it. Also just using `arg.as_ptr()` without `CString::new` doesn't work and results in memory allocation error.

2. Another option might be changing `From` to `TryFrom` then `CString::new(arg)?` would require impl `std::ops::Try` for `TVMValue` which I think is not doable!

So the best solution is to `expect` with the same `CString::new` `NulError` message.

Reply:
I think my real question is why you're using `CString` on an array of bytes that was never actually a C string. Unless I'm confused and this is the actual internal repr that TVM uses.

Reply:
Ah, right! for slice is not necessary just need the clone but for `&str`, etc. is.

Reply:
👍 yep, when the actual type is a `repr C` `&str`, `CString` is definitely appropriate. Otherwise you end up with a potential extra allocation and null terminator byte.